### PR TITLE
[KED-2437] Fix minimum zoom level

### PR DESCRIPTION
**Pull request by bru5**
## Description

Fixes minimum zoom level back to its original behaviour, such that the whole graph can be made visible.

## Development notes

Implemented the fix, some tidying and updated tests.

## QA notes

Zoom out and ensure graph fits fully in view.

Test using different size graphs of different aspect ratios.

Compare to previous release.

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

## Legal notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: src/components/flowchart/flowchart.test.js [hunk 1]
```diff
@@ -88,10 +88,9 @@ describe('FlowChart', () => {
     const instance = wrapper.find('FlowChart').instance();
     const viewExtents = getViewExtents(instance.view);
 
-    expect(viewExtents).toEqual({
-      translate: { minX: -500, minY: -500, maxX: 1044, maxY: 1553 },
-      scale: { minK: 0.8, maxK: 2 },
-    });
+    expect(viewExtents.translate).toEqual({ minX: -500, minY: -500, maxX: 1044, maxY: 1553 });
+    expect(viewExtents.scale.minK).toBeCloseTo(0.777);
```
Comment (studioswong) [inline src/components/flowchart/flowchart.test.js:92]: What is the value `0.777` represent? This would be more intuitive if this is presented as a calculated value rather than an exact figure.

File: src/components/flowchart/flowchart.test.js [hunk 2]
```diff
@@ -108,10 +107,9 @@ describe('FlowChart', () => {
     const instance = wrapper.find('FlowChart').instance();
     const viewExtents = getViewExtents(instance.view);
 
-    expect(viewExtents).toEqual({
-      translate: { minX: -650, minY: -500, maxX: 1479, maxY: 1553 },
-      scale: { minK: 0.8, maxK: 2 },
-    });
+    expect(viewExtents.translate).toEqual({ minX: -650, minY: -500, maxX: 1479, maxY: 1553 });
+    expect(viewExtents.scale.minK).toBeCloseTo(0.777);
```
Comment (studioswong) [inline src/components/flowchart/flowchart.test.js:111]: same as above

File: src/components/flowchart/index.js [hunk 3]
```diff
@@ -283,26 +281,39 @@ export class FlowChart extends Component {
...
+    const minScale = Math.min(
+      chartWidth / (graphWidth || 1),
+      chartHeight / (graphHeight || 1)
+    );
+
     setViewExtents(this.view, {
       translate: {
         minX: -leftSidebarOffset - margin,
-        maxX: width + margin + rightSidebarOffset,
+        maxX: graphWidth + margin + rightSidebarOffset,
```
Comment (studioswong) [inline src/components/flowchart/index.js:311]: We need to add a test to cover the value of `maxX`

File: src/components/flowchart/index.js [hunk 4]
```diff
@@ -283,26 +281,39 @@ export class FlowChart extends Component {
...
         minY: -margin,
-        maxY: height + margin,
+        maxY: graphHeight + margin,
       },
       scale: {
-        minK: this.MIN_SCALE * this.defaultTransform.k,
+        minK: this.MIN_SCALE * minScale,
```
Comment (studioswong) [inline src/components/flowchart/index.js:316]: As with above, we need to add a test to test for the value of `minK`

**Review by richardwestenra — CHANGES_REQUESTED** [review]
Thanks Liam, this is great. It works slightly differently than I'd imagined (I'd thought that it might default to always showing the full graph on first load without the metadata panel open etc), but on further reflection I think it makes sense for it not to work that way and I think I see why you've implemented it this way instead.

I think it'd be worth explaining some of the magic numbers in the tests, but those are pretty minor and I'd consider those optional. The only item of pressing concern is that there are a couple of eslint warnings for unused variables that need tidying up. Thanks!

File: src/components/flowchart/flowchart.test.js (hunk 1 above)
Comment (bru5) [reply src/components/flowchart/flowchart.test.js]: That's fair I should have mentioned these tests were just an integration check for sanity, partly to avoid testing implementation / re-implementing the tested function in the test itself, but I've managed to wrangle it into something fairly reasonable so please check out the updated version.

File: src/components/flowchart/index.js (hunk 3 above)
Comment (bru5) [reply src/components/flowchart/index.js:311]: This should be covered, could you explain further?

File: src/components/flowchart/index.js (hunk 4 above)
Comment (bru5) [reply src/components/flowchart/index.js:316]: This should be covered, could you explain further?

File: src/components/minimap/index.js [hunk 5]
```diff
@@ -31,7 +31,6 @@ export class MiniMap extends Component {
     this.isPointerDown = false;
     this.isPointerInside = false;
     this.lastTransitionTime = 0;
-    this.defaultTransform = origin;
```
[the remark this answers is not in the document]
Comment (bru5) [reply src/components/minimap/index.js:34]: Yeah so I did investigate this in particular for very large / wide graphs, but things ended up getting very complex to handle all the edge cases smoothly. I think we're able to leave that improvement to a later update if needed, now that this fix resolves the minimum zoom.

See above on the tests, I've improved them and fixed the lint warnings.

**Review by richardwestenra — APPROVED** [review]
Nice one, thanks!
